### PR TITLE
Fix bug where size does not work on inputs with a relative path [BA-6359]

### LINKED ISSUE #5477: Size does not work on files in the input section that have a relative path. [BA-6359]
Also posted in JIRA. Posted here for cross-reference by other users.

The size function works correctly for

* absolute paths
* Relative paths from the execution directory.

This means the size function works fine for any outputs. It fails however on inputs of a relative path, since it checks from the execution directory. Since nothing is in the execution directory when evaluating inputs, this always fails.

The size function is context dependent. When calculating outputs, any relative paths should be evaluated from the cromwell execution directory. When calculating the size of inputs any relative paths should be evaluated from the CWD of cromwell (they are also localized that way).

### PR DESCRIPTION
**Pull request by rhpvorderman**
Fixes #5477 

Calculating size of inputs does not work when the inputs are relative. Usually relative paths are calculated from the execution directory. This is fine for outputs, but not for inputs.

This issue was discovered because it is extremely useful to have the size of inputs. Because this way you can guesstimate the resource requirements that should be put in the `runtime` section. Currently it works fine for absolute paths, so that can be used as a workaround.

However: all BioWDL workflows are tested in git repos with inputs JSON files that use relative paths. So we can not test our changes on the CI. Also, some user might want to use relative paths in conjunction with `cromwell run`

This adds a `forInput` boolean to the expressionFunction classes so the size function can be made context-dependent. Currently this is only implemented for the SFSbackend which is presumably the only backend that suffers from this. ~~EDIT: This is now implemented for all backends. It is quite useful as resource estimates using size can be used to request specific VM sizes in the cloud.~~ Size for inputs is now calculated in the same way as inputs with relative paths are localized (from the CWD of the cromwell process).

I added a test that tests this fixed behavior (sizerelativepath). Tests for the current behavior (sizeenginefunction) do not break.

### CODE REVIEW

**Comment by rhpvorderman** [comment]
Hmm it turns out this should be fixed for other backends as well.

**Comment by rhpvorderman** [comment]
One extra clarification about scope here. Although the changes seem quite extensive for this only affects the evaluation of WDL functions at the moment the `input` block in the WDL is processed.  During the `input` block processing, files have not been moved yet so relative paths should be solved naively.

After the `input` block is processed files are localized. The `forInput` Boolean will be set to `false` and relative file paths will be handled differently as required by their backends.

**Comment by rhpvorderman** [comment]
@aednichols Thanks again for allowing me push access to the repo. I can not test all the backends manually so it is good that I can access the CI environment and see if the bug fix turned out well.

What is the formal process of getting this bug under the Cromwell team's attention? I have made a JIRA issue. Should I put it on the sprint? I also ask this for #5456 which is a really simple fix. I am not in great haste getting a review, but I want to ensure these fixes end up in the next release of Cromwell. These bugs are now actively blocking BioWDL development as our CI always uses a mainline version of Cromwell. (Usually the latest, but we are already actively excluding 49 because of the relative outputs bug). 

By no means I want to push the Cromwell team in reviewing these fixes right now, but if you could give me some procedure that would make sure these are reviewed before the next release is out, that would give me some peace of mind. Thanks!

**Comment by aednichols** [comment]
@rhpvorderman we only have the bandwidth to review a couple PRs at a time. We mark those with the label "on deck for review".

**Comment by rhpvorderman** [comment]
@aednichols Thanks for the info. I will just be patient then :slightly_smiling_face: .

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,12 @@
 # Cromwell Change Log
 
+## 50 Release Notes
+
+### Bug fixes
+
+* Fixed a bug where the `size(...)` function did not work correctly on files from the input section
+  of a task that had a relative path.
+
```
Comment (aednichols) [inline CHANGELOG.md:9]: Sorry... have to relocate these to 51 also

**Comment by cjllanwarne** [comment]
This fixes the problem at the point of expression evaluation... it seems like it might be easier (and a lot less fiddly?) to do the relative file resolution much earlier, at the point that inputs are being read in to the workflow in the first place.

The `ValidatedWomNamespace` produced as part of workflow materialization contains a `womValueInputs` field... I wonder whether performing this mapping as part of creating that validated set of inputs would work?

**Comment by rhpvorderman** [comment]
> I can follow how `forInput` is passed around, but I can't seem to discern where it is actually set - i.e. where the default value is overridden. Can you help me wrap my head around this?

[Here](https://github.com/broadinstitute/cromwell/blob/9bd4e90fdf999abc76d0ba8e801ad24eb99140b5/engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/job/preparation/JobPreparationActor.scala#L62) in the JobPreparationActor.

**Comment by rhpvorderman** [comment]
> This fixes the problem at the point of expression evaluation... it seems like it might be easier (and a lot less fiddly?) to do the relative file resolution much earlier, at the point that inputs are being read in to the workflow in the first place.

> The ValidatedWomNamespace produced as part of workflow materialization contains a womValueInputs field... I wonder whether performing this mapping as part of creating that validated set of inputs would work?

Great suggestion. I will take a look at this. I can checkout the test case on a new branch and try to hack there. One of the catches will be that this resolving will be backend dependent.

In the current situation the input expressions are evaluated first, and after that the inputs are resolved. (This makes sense because input can also be something like `baseDir + "/my_file.txt"`, which needs to be evaluated).

But indeed this could be bypassed by doing this already at the workflow level, before it gets passed down to the task level. I will take a look at this. If it does not work, (or work easily) then I will report back here.

**Comment by rhpvorderman** [comment]
@cjllanwarne I have had a look bit it is quite non-trivial to find how to evaluate paths at the womValueInputs stage, as cromwell has no knowledge of the backend at this point. This "relative path" business only makes sense for the sfsBackend which handles access to local systems. For cloud systems relative paths make no sense at all. 
Evaluating it at the JobPreparationActor is the correct spot, although the implementation ends up to be a bit messy :confused: .

**Comment by rhpvorderman** [comment]
@aednichols @cjllanwarne I made a copy of this PR with the same tests and changelog, but with a less intrusive code change at #5495 .

**Comment by rhpvorderman** [comment]
Closing this as the solution in #5495 has now grown to become more elegant and less intrusive.

**Pull request closed without merging**